madminer/utils/morphing.py

```diff
@@ -385,44 +645,114 @@ def calculate_morphing_weights(self, theta, basis=None, morphing_matrix=None):
+            component_weights = self._calculate_component_weight(theta=theta)
+
+            # Transform to basis weights
+            return morphing_matrix.T.dot(component_weights)
+        else:  # calculate matrix with gd, gp, gs
+            if gs is None:
+                gs = self.gs
+            if gd is None:
+                gd = self.gd
+            if gp is None:
+                gp = self.gp
+
+            if gs is None and gd is None and gp is None:
+                raise RuntimeError(
+                    "No basis defined or given. Use PhysicsMorpher.set_basis(), PhysicsMorpher.optimize_basis(), or the "
+                    "basis keyword."
+                )
 
-        if basis is None:
-            raise RuntimeError(
-                "No basis defined or given. Use PhysicsMorpher.set_basis(), PhysicsMorpher.optimize_basis(), or the "
-                "basis keyword."
-            )
+            if morphing_matrix is None:
+                morphing_matrix = self.calculate_morphing_matrix(basis, gs, gd, gp)
 
-        if morphing_matrix is None:
-            morphing_matrix = self.calculate_morphing_matrix(basis)
+            if theta_d is None and theta_p is None and theta_s is None:
+                # calculate component weights with theta
+                component_weights = self._calculate_component_weight(theta)
 
-        # Calculate component weights
-        component_weights = np.zeros(self.n_components)
-        for c in range(self.n_components):
-            factor = 1.0
-            for p in range(self.n_parameters):
-                factor *= float(theta[p] ** self.components[c, p])
-            component_weights[c] = factor
-        component_weights = np.array(component_weights)
+            else:  # calculate component weights with theta_d, theta_p and theta_s
+                component_weights = np.zeros(self.n_components)
+                n_s = 0
+                n_p = 0
+                n_d = 0
 
-        # Transform to basis weights
-        return morphing_matrix.T.dot(component_weights)
+                if theta_s is not None:
+                    n_s = len(theta_s)
+                if theta_p is not None:
+                    n_p = len(theta_p)
+                if theta_d is not None:
+                    n_d = len(theta_d)
 
-    def calculate_morphing_weight_gradient(self, theta, basis=None, morphing_matrix=None):
+                for c in range(self.n_components):
+                    factor = 1.0
+                    if n_d != 0:
+                        for j in range(n_d):
+                            factor *= float(theta_d[j] ** self.components[c, j])
+                    if n_p != 0:
+                        for i in range(n_p):
+                            if n_d != 0:
+                                factor *= float(theta_p[i] ** self.components[c, i + n_d])
+                            else:
+                                factor *= float(theta_p[i] ** self.components[c, i])
```

Comment:
This `if` - `else` block can be simplified. If you are adding the value `n_d` to `i` only when it is not zero, you might as well do it always, even if it is zero (_innocuous operation_) 🤷🏻‍♂️

```python
factor *= float(theta_p[i] ** self.components[c, i + n_d])
```

Reply:
Thank you for your feedback! I made this change and the below one in the new commit